src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -799,23 +998,19 @@ public PrefixEnumerator(ModelStateDictionary dictionary, string prefix)
                     throw new ArgumentNullException(nameof(prefix));
                 }
 
-                _dictionary = dictionary;
-                _prefix = prefix;
-
-                _exactMatch = ExactMatchState.NotChecked;
-                _enumerator = default(Dictionary<string, ModelStateEntry>.Enumerator);
-                Current = default(KeyValuePair<string, ModelStateEntry>);
+                _index = 0;
+                _rootNode = dictionary.FindNode(prefix);
+                _nodes = new List<EnumeratorEntry>
```

Comment:
![image](https://cloud.githubusercontent.com/assets/1430011/13119652/289f437e-d55f-11e5-8341-94d6a7609ff9.png)

Are you doing this to maintain ordering?